wallet/src/key_chain/leaf_key_chain.rs

```diff
@@ -0,0 +1,517 @@
+                id.clone(),
+                KeyPurpose::Change,
+                account_pubkey
+                    .clone()
+                    .derive_child(KeyPurpose::Change.get_deterministic_index())?,
+                change_addresses,
+                change_public_keys,
+                usage_states
+                    .remove(&KeyPurpose::Change)
+                    .ok_or(KeyChainError::MissingDatabaseProperty("Change usage state"))?,
+            )?,
+        ))
+    }
+
+    pub fn issue_address<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        lookahead_size: u32,
+    ) -> KeyChainResult<Address> {
+        let new_issued_index = self.get_new_issued_index(lookahead_size)?;
+
+        // Get address or derive one if necessary
+        let issued_address = match self.addresses.get(&new_issued_index) {
+            Some(address) => address.clone(),
+            None => {
+                self.derive_and_add_key(db_tx, new_issued_index)?;
+                self.addresses
+                    .get(&new_issued_index)
+                    .expect("The address should be derived")
+                    .clone()
+            }
+        };
+
+        self.set_key_index_as_issued(db_tx, new_issued_index)?;
+
+        Ok(issued_address)
+    }
+
+    /// Issue a new key. This does not check if lookahead margins are observed
+    pub fn issue_key<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        lookahead_size: u32,
+    ) -> KeyChainResult<ExtendedPublicKey> {
+        let new_issued_index = self.get_new_issued_index(lookahead_size)?;
+
+        // Get key or derive one if necessary
+        let issued_key = match self.derived_public_keys.get(&new_issued_index) {
+            Some(key) => key.clone(),
+            None => self.derive_and_add_key(db_tx, new_issued_index)?,
+        };
+
+        self.set_key_index_as_issued(db_tx, new_issued_index)?;
+
+        Ok(issued_key)
+    }
+
+    /// Set a specific key index as used
+    fn set_key_index_as_issued<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        new_issued_index: ChildNumber,
+    ) -> KeyChainResult<()> {
+        // Save usage state
+        self.usage_state.set_last_issued(Some(new_issued_index));
+        self.save_usage_state(db_tx)
+    }
+
+    /// Persist the usage state to the database
+    pub fn save_usage_state<B: Backend>(&self, db_tx: &mut StoreTxRw<B>) -> KeyChainResult<()> {
+        Ok(db_tx.set_keychain_usage_state(
+            &AccountKeyPurposeId::new(self.account_id.clone(), self.purpose),
+            &self.usage_state,
+        )?)
+    }
+
+    /// Get a new issued index and check that it is a valid one i.e. not exceeding the lookahead
+    fn get_new_issued_index(&self, lookahead_size: u32) -> KeyChainResult<ChildNumber> {
+        // TODO consider last used index as well
+        let new_issued_index = {
+            match self.usage_state.get_last_issued() {
```

Comment:
Again, you can use `self.get_last_issued()` and remove the dead code annotation.